scrapy/core/downloader/handlers/s3.py

```diff
@@ -65,12 +70,30 @@ def download_request(self, request, spider):
         bucket = p.hostname
         path = p.path + '?' + p.query if p.query else p.path
         url = '%s://%s.s3.amazonaws.com%s' % (scheme, bucket, path)
+        # boto headers are different from scrapy headers:
+        # they expect unicode keys and values in python 3 and do not handle
+        # multiple values.
+        headers_to_sign = CaselessDict()
+        for key, values in request.headers.items():
+            value = b','.join(values)
```

Comment:
It could make sense to make this an utility function (convert to unicode, join values), or even a Headers method.